lcls_tools/common/data/emittance/matrix_calcs.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+
+
+def propagate_twiss(twiss_init: np.ndarray, rmat: np.ndarray):
+    """
+    Propagates twiss parameters downstream given a transport rmat.
+
+    Parameters:
+        twiss_init: numpy array shape batchshape x 3 containing the initial twiss params
+                    (ordered: beta, alpha, gamma)
+        rmat: numpy array shape batchshape x 2 x 2 containing 2x2 transport rmats
+
+    Outputs:
+        twiss_final: numpy array shape batchshape x 3 containing the downstream twiss params
+                    (ordered: beta, alpha, gamma)
+    """
+    twiss_transport = twiss_transport_mat_from_rmat(
+        rmat
+    )  # result shape (batchshape x 3 x 3)
+
+    twiss_final = twiss_transport @ np.expand_dims(twiss_init, axis=-1)
+    # result shape (batchshape x 3 x 1)
+
+    return twiss_final.squeeze(-1)
+
+
+def twiss_transport_mat_from_rmat(rmat: np.ndarray):
+    """
+    Converts from 2x2 rmats to 3x3 twiss transport matrices.
+
+    Parameters:
+        rmat: numpy array shape batchshape x 2 x 2
+
+    Outputs:
+        result: numpy array shape batchshape x 3 x 3
+    """
+    # Converts from 2x2 rmats to 3x3 twiss transport matrices.
+    c, s, cp, sp = rmat[..., 0, 0], rmat[..., 0, 1], rmat[..., 1, 0], rmat[..., 1, 1]
+    result = np.stack(
+        (
+            np.stack((c**2, -2 * c * s, s**2), axis=-1),
+            np.stack((-c * cp, c * sp + cp * s, -s * sp), axis=-1),
+            np.stack((cp**2, -2 * cp * sp, sp**2), axis=-1),
+        ),
+        axis=-2,
+    )  # result shape (batchshape, 3, 3)
+    return result
+
+
+def build_quad_rmat(k: np.ndarray, q_len: float, thin_lens: bool=False):
+    """
```

Comment:
Can some of of these rmats be pulled from the control system directly?

Reply:
Maybe, I'm not sure. In cases where computation time is a factor, I think it would be preferable to not need to make adjustments to the model/retrieve the rmats from it. The idea for this code is to be able to handle large batches of scan data very efficiently.

Reply:
There may be a way to grab and store this info vs. calling it new every time. Let's open an issue for that linked to this PR, and we can merge this for now. I'm more worried about differences in rmats in the models used for other applications.